### ISSUE TITLE
Merge bot using GitHub Actions

### ISSUE THREAD

**Issue by marcelosousa** [reviewpad/reviewpad#391]
A merge bot is useful to guarantee that a production branch respects certain conditions (the most important being that the CI is green).

This is specially important for teams that continuously deploy to production on every push to such branches and also have a high load of pull requests.

We can speculate that in fact most users of Mergify do it because of the merge bot feature.

GitHub is also rolling their Merge Queue feature https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta.

There are other players in this market such as https://www.aviator.co.

The feedback that I got regarding the GitHub merge bot from an user was that:

1. Pull requests in the merge queue run the CI again even if the CI was already green (the time and resource consumption doubles which is very annoying).
2. GitHub already tries to speculate by combining multiple pull requests in one go (they don't like this because it prevents them from having traceability of production errors - they prefer to wait longer and getting the sequential merges)
3. One very nice aspect of GitHub merge queues is that it predicts the time that will take until the pull request is merged which seems to be quite accurate. (Since they have the timing information of the CI pipelines they just have to average it out and then multiply by the number of elements in the queue).

----

As a first baby step what we would like to implement is a merge bot with the following inputs:
1. trigger (e.g. a comment `/reviewpad merge` or a particular label) 
2. merge criteria 

As soon as trigger occurs, the pull request is added to the queue.

Then, Reviewpad will take the first element of the queue and check if it is up to the date with the base branch. If it is and the merge criteria is true, merges. If it is not up to date, then perform a rebase operation. If it is up to date and the merge criteria does not apply or if it has git conflicts, remove the from the merge queue with a notification.

Regarding the implementation of a merge bot/queue by Reviewpad this is significant challenge just using GitHub actions. The main reason is that we can't keep a state solely using GitHub actions and Reviewpad typically runs in the context of a single pull request.

This means that if two pull requests were added to the merge queue in a very short period of time, if the code that picks the head of the queue is ran in parallel it could pick two different pull requests. The only solution that we found so far with GitHub Actions would be use a cron job. That is, every 10 minutes the merge queue job will run regardless of the state of the queue. 

However, adding a cron job means that we need a separate GitHub action because the current Reviewpad architecture is really made to be executed on a single pull request.

In practice, we can't combine the specification of a merge queue with the rest of the workflows. FYI: The same happens with the stale pull request check.

For example, suppose that you had a workflow:

```
workflows:
   - name: always-add-a-triage-label
      if: true
      then:
         - $addLabel("triage")
```

If we run this workflow together with the merge queue, a label triage would be added every 10 minutes even if the user had already removed such label.

### Technical specification of the merge bot

The interaction starts with the command `/reviewpad merge method`  

- Actions to take:
    1. Comment PR with feedback message saying `Pull request added to the merge queue. Now set and relax.`
    2. Add the label `in-merge-queue`
    3. Checks if pr is up to date.
        1. No - Do `%CODE_1%`
        2. Yes - Do nothing

The next step is taking care by a cron job. Specifications:

- The cron runs every 10 minutes and:
    
    ```go
    foreach pr in pr_with_label("in-merge-queue")
         if isUpToDate(pr) {
    				if mergeCondition(pr) {
    					merge(pr, method) // merge into base with selected method
            } else {
    				  // do nothing
    					// the cron job will run again on the next 10 minutes
            }
    	   } else {
            // %CODE_1%
    				if hasGitConflicts(pr) {
    					removeLabel("in-merge-queue")
    					addLabel("git-conflicts") // optional
    					comment("this pr was removed from merge queue due to git conflicts")
    			  } else {
    					rebase(pr)
            }
    		 }
    
    // where pr_with_label selects the pr with the give label by order in which the label was added
    ```
    

reviewpad.yml example

```yaml
api-version: reviewpad.com/v3.x

rules:
  - name: mergeable
    spec: $workflowStatus("ci") == green	

commands:
	merge: # /reviewpad merge squash
		parameters:
			- method: string
    ## I THINK THIS IS THE PROBLEM
		action: $workflowStatus("ci") == green
	label: # /reviewpad label outdated
		parameters:
			- name: string
		action: $addLabel($name)

workflow:
  - name: comment-every-twenty-minutes
    if: true
    then:
      - $comment("Another 20 minutes")
```

github action example

```yaml
name: Reviewpad Action

on:
  issue_comment:
  schedule:
    # Run every 20 minutes
    - cron: "0 0 * * *"
    # Run for the merge bot
    - cron: "0 0 * * *"

jobs:
  reviewpad-action:
    runs-on: ubuntu-latest
    steps:
      - name: Running reviewpad action
        uses: reviewpad/action@v3.x
```

### **Issues**:

- Using a cron job makes reviewpad to be **unpredictable**.
- Make **all-in-one** solution makes it hard when we want workflows to be triggered on certain conditions, for instance,
    
    ```yaml
    workflow:
    	# Run workflow every 20 minutes (perfect for cron job)
      - name: comment-every-20-minutes
    		if: "true"
        then:
          - $comment("Another 20 minutes")
      # This workflows does not make sense to run every 20 minutes
      - name: welcome
        if:
    			- rule: $totalCreatedPullRequests($author) == 1
        then:
          - $commentOnce("Welcome")
    ```
    

### Current state

1. GitHub Action is triggered `on` event
2. Calls Reviewpad
3. Reviewpad handles to `processor` which picks the correct pull requests/issues to run the reviewpad on - based on the triggered `event`
4. Then foreach pull request/issue
    1. Run reviewpad

Problems:

- cron jobs
- merge bot

### MERGE BOT

action

```yaml
name: Reviewpad Action

on:
  schedule:
    # Run every 10 minutes
    - cron: "10 * * * *"

jobs:
  reviewpad-action:
    runs-on: ubuntu-latest
    steps:
      - name: Running reviewpad action
        uses: reviewpad/action@v3.x
```

reviewpad.yml

```yaml
api-version: reviewpad.com/v3.x

rules:
  - name: mergeable
    spec: $workflowStatus("ci") == green

mergebot:	
	condition: $rule("mergeable")
```

- WITHOUT CRON JOB
    1. User comment `/reviewpad merge squash`
    2. Trigger reviewpad to run
        
        ```go
        // Handle github action event
        if event = comment {
        	if comment.startsWith("/reviewpad") {
        		if command == merge {
        			// Merge bot kicks in
            } else {
        			// Add support to reviewpad commands?
            }
        	}
        }
        
        if event = label("in merge queue") {
        	sleep(30) //to make sure head is already selected - DOES NOT WORK
        	head_pr = head_of_merge_queue()
          if merge_condition(head_pr) {
        		merge(head_pr)
        		remove_label("head")
        		remove_label("in merge queue")
        		trigger_new_merge_bot_run_in(0)
        	} else {
        		trigger_new_merge_bot_run_in(10)
        	}
        }
        ```
        
        ```go
        merge_bot(pr) {
        	if git_conflicts(pr) {
        	  comment("pr with git conflicts")
        	} else {
        		add_label("in merge queue")
        		comment("pr added to the queue")
        		if !is_outdated(pr) {
        			rebase(pr)
        		}
        	}
        }
        ```
        
        ```go
        head_of_merge_queue() {
        	head = get_pr_with_label("head")
        	if head {
        		return
        	}
        	prs  = get_all_prs_with_label("in merge queue")
        	head = get_pr_with_oldest_command(prs, "comment")
        	add_label(pr, "head")
        	return head
        }
        ```
        
        ```go
        trigger_new_merge_bot_run_in(minutes) {
        	// HOW TO HANDLE THE MINUTES? (sleep?) 🤬
        	// ALSO REQUIRES GITHUB TOKEN TO TRIGGER ANOTHER WORKFLOW 🤬
        	// Use GitHub API
        	// /repos/{owner}/{repo}/actions/workflows/{workflow_id}/dispatches
        }
        ```
        
- WITH CRON JOB
    
    ```yaml
    name: Reviewpad Merge Bot
    
    on:
      schedule:
        # Run every 10 minutes
        - cron: "10 * * * *"
    
    jobs:
      reviewpad-merge-bot:
        runs-on: ubuntu-latest
        steps:
          - name: Running reviewpad action
            uses: reviewpad/mergebot@v1
    				with:
    					- condition: $workflowStatus("ci") == green
    ```
    
    ```bash
    /reviewpad merge squash
    ```
